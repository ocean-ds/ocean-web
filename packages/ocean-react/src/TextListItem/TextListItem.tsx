import { ChevronRight } from '@useblu/ocean-icons-react';
import classNames from 'classnames';
import React, { useEffect, useMemo, useRef, useState } from 'react';
import Tag from '../Tag';
import Checkbox, { CheckboxProps } from '../Checkbox/Checkbox';
import Radio, { RadioProps } from '../Radio/Radio';

export type TextListItemProps = {
  /**
   * The title of the component.
   */
  title: string;
  /**
   * The description of the component.
   */
  description: string;
  /**
   * The caption of the component.
   */
  caption?: string;
  /**
   * Show an action icon on the right side of the component.
   */
  tagLabel?: string;
  /**
   * The info text of the component.
   */
  infoText?: string;
  /**
   * The type of the info text.
   */
  infoTextType?: 'neutral' | 'positive';
  /**
   * Show an action icon on the right side of the component.
   */
  withAction?: boolean;
  /**
   * Callback function when the action icon is clicked.
   */
  onActionClick?: () => void;
  /**
   * Show a checkbox on the left side of the component.
   */
  checkbox?: CheckboxProps;
  /**
   * Show a radio on the left side of the component.
   */
  radio?: RadioProps;
  /**
   * The class name of the component.
   */
  className?: string;
} & React.ComponentPropsWithoutRef<'div'>;

const TextListItem = React.forwardRef<HTMLDivElement, TextListItemProps>(
  (
    {
      title,
      description,
      caption,
      tagLabel,
      infoText,
      infoTextType = 'neutral',
      withAction,
      onActionClick,
      checkbox,
      radio,
      className,
      ...rest
    },
    ref
  ) => {
    const buttonElementRef = useRef<HTMLButtonElement | null>(null);
    const divElementRef = useRef<HTMLDivElement | null>(null);
    const [isWide, setIsWide] = useState(false);

    useEffect(() => {
      const updateWidth = () => {
        if (buttonElementRef.current) {
          const width = buttonElementRef.current.offsetWidth;
          setIsWide(width >= 992);
        }
        if (divElementRef.current) {
          const width = divElementRef.current.offsetWidth;
          setIsWide(width >= 992);
        }
      };

      updateWidth();
      window.addEventListener('resize', updateWidth);

      return () => {
        window.removeEventListener('resize', updateWidth);
      };
    }, []);

    const textListitem = useMemo(
      () => (
        <button
          ref={buttonElementRef}
          onClick={onActionClick}
          type="button"
          className="ods-text-list-item--button"
        >
          <div
            className={classNames('ods-text-list-item', className, {
              'ods-text-list-item--with-action': withAction,
              'ods-text-list-item--wide': isWide,
              'ods-text-list-item--selectable': checkbox || radio,
            })}
            ref={ref}
            {...rest}
          >
            <div className="ods-text-list-item__default-content">
              <p className="ods-typography ods-typography__paragraph">
                {title}
                {tagLabel && (
                  <Tag type="neutral" size="medium">
                    {tagLabel}
                  </Tag>
                )}
              </p>
              <p className="ods-typography ods-typography__description">
                {description}
              </p>
              {caption && (
                <p className="ods-typography ods-typography__captionbold">
                  {caption}
                </p>
              )}
            </div>
            {infoText && (
              <p
                className={classNames(
                  'ods-typography ods-typography__description',
                  'ods-text-list-item__info-text',
                  `ods-text-list-item__info-text--${infoTextType}`
                )}
              >
                {infoText}
              </p>
            )}
            {withAction && <ChevronRight />}
          </div>
        </button>
      ),
      [
        title,
        description,
        caption,
        tagLabel,
        infoText,
        infoTextType,
        withAction,
        onActionClick,
        checkbox,
        radio,
        className,
        isWide,
        ref,
        rest,
      ]
    );

    if (checkbox) {
      return (
        <div className="ods-text-list-item-selectable" ref={divElementRef}>
          <Checkbox {...checkbox} label={textListitem} />
        </div>
      );
    }

    if (radio) {
      return (
        <div className="ods-text-list-item-selectable" ref={divElementRef}>
          <Radio {...radio} label={textListitem} />
        </div>
      );
    }

    return textListitem;
  }
);

TextListItem.displayName = 'TextListItem';

export default TextListItem;
