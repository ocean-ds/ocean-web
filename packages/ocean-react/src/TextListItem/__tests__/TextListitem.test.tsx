import React from 'react';
import { render, screen } from '@testing-library/react';
import TextListItem, { TextListItemProps } from '../TextListItem';

describe('TextListItem', () => {
  const defaultProps: TextListItemProps = {
    title: 'Sample Title',
    description: 'Sample Description',
    caption: 'Sample Caption',
    tagLabel: 'New',
    infoText: 'Info Text',
    infoTextType: 'neutral',
    withAction: true,
    onActionClick: jest.fn(),
    className: 'custom-class',
  };

  test('should render with action and match snapshot', () => {
    const { asFragment } = render(
      <TextListItem
        title={defaultProps.title}
        description={defaultProps.description}
        caption={defaultProps.caption}
        withAction={defaultProps.withAction}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('should render with info text and match snapshot', () => {
    const { asFragment } = render(
      <TextListItem
        title={defaultProps.title}
        description={defaultProps.description}
        infoText={defaultProps.infoText}
        infoTextType={defaultProps.infoTextType}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('should render with a checkbox if the checkbox prop is provided', () => {
    const { asFragment } = render(
      <TextListItem
        title={defaultProps.title}
        description={defaultProps.description}
        checkbox={{ checked: false }}
      />
    );
    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByText('Sample Title')).toBeInTheDocument();
  });

  test('should render with a radio if the radio prop is provided', () => {
    const { asFragment } = render(
      <TextListItem
        title={defaultProps.title}
        description={defaultProps.description}
        radio={{ checked: true }}
      />
    );
    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByText('Sample Title')).toBeInTheDocument();
  });
});
