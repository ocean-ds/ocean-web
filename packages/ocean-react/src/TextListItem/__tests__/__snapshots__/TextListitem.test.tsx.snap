// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextListItem should render with a checkbox if the checkbox prop is provided 1`] = `
<DocumentFragment>
  <div
    class="ods-text-list-item-selectable"
  >
    <div
      class="ods-checkbox__root-container"
    >
      <label
        class="ods-checkbox__root"
      >
        <input
          class="ods-checkbox"
          type="checkbox"
        />
        <span
          class="ods-checkbox__checkmark"
        />
        <span
          class="ods-typography ods-typography__description ods-checkbox__label"
        >
          <button
            class="ods-text-list-item--button"
            type="button"
          >
            <div
              class="ods-text-list-item ods-text-list-item--selectable"
            >
              <div
                class="ods-text-list-item__default-content"
              >
                <p
                  class="ods-typography ods-typography__paragraph"
                >
                  Sample Title
                </p>
                <p
                  class="ods-typography ods-typography__description"
                >
                  Sample Description
                </p>
              </div>
            </div>
          </button>
        </span>
      </label>
    </div>
  </div>
</DocumentFragment>
`;

exports[`TextListItem should render with a radio if the radio prop is provided 1`] = `
<DocumentFragment>
  <div
    class="ods-text-list-item-selectable"
  >
    <label
      class="ods-radio__root"
    >
      <input
        checked=""
        class="ods-radio"
        type="radio"
      />
      <span
        class="ods-radio__checkmark"
      />
      <span
        class="ods-typography ods-typography__description ods-radio__label"
      >
        <button
          class="ods-text-list-item--button"
          type="button"
        >
          <div
            class="ods-text-list-item ods-text-list-item--selectable"
          >
            <div
              class="ods-text-list-item__default-content"
            >
              <p
                class="ods-typography ods-typography__paragraph"
              >
                Sample Title
              </p>
              <p
                class="ods-typography ods-typography__description"
              >
                Sample Description
              </p>
            </div>
          </div>
        </button>
      </span>
    </label>
  </div>
</DocumentFragment>
`;

exports[`TextListItem should render with action and match snapshot 1`] = `
<DocumentFragment>
  <button
    class="ods-text-list-item--button"
    type="button"
  >
    <div
      class="ods-text-list-item ods-text-list-item--with-action"
    >
      <div
        class="ods-text-list-item__default-content"
      >
        <p
          class="ods-typography ods-typography__paragraph"
        >
          Sample Title
        </p>
        <p
          class="ods-typography ods-typography__description"
        >
          Sample Description
        </p>
        <p
          class="ods-typography ods-typography__captionbold"
        >
          Sample Caption
        </p>
      </div>
      <svg
        fill="currentColor"
        height="24"
        viewBox="0 0 20 20"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M7.293 14.707a1 1 0 0 1 0-1.414L10.586 10 7.293 6.707a1 1 0 0 1 1.414-1.414l4 4a1 1 0 0 1 0 1.414l-4 4a1 1 0 0 1-1.414 0z"
          fill-rule="evenodd"
        />
      </svg>
    </div>
  </button>
</DocumentFragment>
`;

exports[`TextListItem should render with info text and match snapshot 1`] = `
<DocumentFragment>
  <button
    class="ods-text-list-item--button"
    type="button"
  >
    <div
      class="ods-text-list-item"
    >
      <div
        class="ods-text-list-item__default-content"
      >
        <p
          class="ods-typography ods-typography__paragraph"
        >
          Sample Title
        </p>
        <p
          class="ods-typography ods-typography__description"
        >
          Sample Description
        </p>
      </div>
      <p
        class="ods-typography ods-typography__description ods-text-list-item__info-text ods-text-list-item__info-text--neutral"
      >
        Info Text
      </p>
    </div>
  </button>
</DocumentFragment>
`;
